Move code actions to LSP

allisonchou opened the pull request:
Moves code actions over to LSP.

How it works:
There are three main handlers involved in processing LSP code actions: `CodeActionsHandler`, `CodeActionResolveHandler`, and `RunCodeActionsHandler`. 

1. `CodeActionsHandler`: Deals with the initial code actions request from LSP. Returns an array of `VSCodeActions` generated from Roslyn code actions. The `Edit` and `Command` properties in the `VSCodeAction` are left blank, since those are populated in the `CodeActionResolveHandler`. We avoid filling out those two properties so we don't need to a bunch of work unless the user actually requests it.
2. `CodeActionResolveHandler`: Invoked when a user hovers over a code action. We leave the bulk of the processing to this handler. Fills out either the `Edit` or `Command` properties (can be both as well, but due to the LSP bug below this isn't currently possible). If we send back an `Edit`, the edit is processed and executed on the client/LSP side. If we send back a `Command`, the command will eventually make its way to the `RunCodeActionsHandler` and run from there.
3. `RunCodeActionsHandler`: Invoked when `Command` property is filled out. We run actions as commands if they can't be executed as WorkspaceEdits for some reason (i.e. if not an ApplyChangesOperation, or due to the LSP bug below). These commands are executed on the local Roslyn side. The general `ExecuteWorkspaceCommandHandler` is called first. The general handler then calls into our code-action specific `RunCodeActionsHandler`.

Code actions that add a document must currently be executed on the server side due to this LSP bug: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/

eta: thanks to David for all the help on this PR!

allisonchou commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,54 +38,90 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+                var configureSuppressActions = codeActions.Where(a => a.Key is AbstractConfigurationActionWithNestedActions);
+                if (configureSuppressActions.Any())
+                {
+                    results.Add(GenerateVSCodeAction(request, new CodeActionWithNestedActions(
```
The configure and suppress issues is a code action that is dealt with specially in local Roslyn.
We can't reuse the resource string since it lies in a different layer.

dibarbet replied:
Can we move the existing string rather than duplicate?

allisonchou replied:
I think we could, although the original location makes more sense to me for the local scenario.
@mavasani, is there any issue with moving the "Configure or Suppress issues" string to CodeFixesResources instead of EditorFeaturesWpfResources so that we can use it in the LSP layer?

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveData.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    internal class CodeActionResolveData
+    {
+        public CodeActionParams CodeActionParams { get; set; }
```
rather than pass in the entire parameters, lets pass in just what we need

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,279 @@
+            CancellationToken cancellationToken)
+        {
+            CodeActionResolveData data;
+            if (codeAction.Data is CodeActionResolveData codeActionResolveData)
```
this was due to liveshare previously deserializing things for us, I believe it should only ever be jtoken now that it's coming through the LSP client.

allisonchou replied:
I think we need to keep in the first check for the code action resolver tests. I can make a comment to note that. I'm not sure if there's a workaround?

dibarbet replied:
I think in the tests we can serialize it to string before putting it in the data field.  would be better than keeping this check around I think

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,279 @@
+            var runAsCommand = false;
+
+            var applyChangesOperations = operations.Where(operation => operation is ApplyChangesOperation);
+            if (applyChangesOperations.Any())
```
I wonder if this should be all, or more specific filtering.

If we have an ApplyChangesOperation + DocumentNavigation operation, then it should be safe to do this (we can't tell the client to navigate anywhere anyway, so we can't do anything but drop the document navigation operation).  But if there is an ApplyChangesOperation + some other actual modification operation, then we might get into a corrupt state b/c we applied half of the code action.

Not sure the best way to do this, thoughts @CyrusNajmabadi ?

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,279 @@
+            }
+
+            // Set up to run as command on the server instead of using WorkspaceEdits.
+            var commandOperations = operations.All(operation => !(operation is ApplyChangesOperation));
```
Ah I see, we apply any non ApplyChangesOperation as a command.  Then I think I'm only worried about the order.  The LSP spec states that commands are applied after workspace edits.  But if the code action specifies the non-ApplyChangesOperation to run before the ApplyChangesOperation, the code action could get corrupted if order matters.

dibarbet replied:
This also could be problematic - I see in the execute code actions request, we just find the one with the right title.  It will end up executing all of the operations, so if there are workspace edits + a command, it will execute the edits twice (once on client workspace edits, once on server during RunCodeActions ApplyChangesOperation)

allisonchou replied:
Do you think it would be best if we were to just file a Roslyn bug on this?
I think ideally, the solution would be to include a check in RunCodeActionsHandler to see whether we are executing an ApplyChangesOperation so that we don't execute those kinds of operations twice. But we can't currently do that since some ApplyChangesOperations are applied as commands due to [this](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/) bug on the LSP side. Once the LSP bug is fixed, I think we can address the concerns you mentioned.

The current approach is: if we have a mix of ApplyChangesOperations and other operations, only the ApplyChangesOperations are executed. But if we have only non-ApplyChangesOperations, then we can execute them (as a command). This isn't ideal, but it prevents workspace edits from potentially executing twice.

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,279 @@
+
+                    // TO-DO: If the change involves modifying any document besides the document where the code action
+                    // was invoked, temporarily execute via command instead of WorkspaceEdit until LSP bug is fixed:
+                    // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
```
So many lsp bugs :(

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,279 @@
+
+                    var changedDocuments = projectChanges.SelectMany(pc => pc.GetChangedDocuments());
+                    var changedAnalyzerConfigDocuments = projectChanges.SelectMany(pc => pc.GetChangedAnalyzerConfigDocuments());
+                    var changedAdditionalDocuments = projectChanges.SelectMany(pc => pc.GetChangedAdditionalDocuments());
```
@CyrusNajmabadi Is there a way to specify get all diffs between two solutions, including analyzer and additional documents?  Rather than having to go document by document?

CyrusNajmabadi replied:
not that im' aware of.

CyrusNajmabadi replied:
also, we are specially handling adding a doc, but not removing?  the approach here is very centered about supporting our code actions.  but would nto necessarily work for other code actions (or if we add code actinos in the future that do more than add-docs or edit docs.

CyrusNajmabadi replied:
my concern still applies here.  i'm not sure what the philosophy is, but it seems like we're not handling the full set of workspace edits that would normally be possible.  that makes me uneasy and makes me feel like we could easily ship broken stuff.  i.e. someone writes a feature like "move type" and expects to be able to remove a file.  We write it and test it and validate it in VS.  but then in LSP itis just broken and we wouldn't know until someone hits it.

i do not like things like that because it just means different parties have to know all the implemented/unimplemnted capabilities of all these subcomponents.

allisonchou replied:
I added code so that any code action that executes as a command should (for now) be executed as a command.

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,54 +38,90 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+
+                // Special case (also dealt with specially in local Roslyn): 
+                // If we have configure/suppress code actions, combine them under one top-level code action.
+                var configureSuppressActions = codeActions.Where(a => a.Key is AbstractConfigurationActionWithNestedActions);
```
are these not specified as NestCodeActions in the CodeAction?  Seems a little weird we have to handle these separately.

allisonchou replied:
Yes, they aren't specified as nested code actions - Roslyn handles them specially [here](http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.Wpf/Suggestions/SuggestedActionsSource.cs,683).

dibarbet commented on src/VisualStudio/LiveShare/Impl/PreviewCodeActionsHandler.cs:
```diff
@@ -42,7 +41,7 @@ public async Task<LSP.TextEdit[]> HandleAsync(RunCodeActionParams request, Reque
                 request.CodeActionParams.Range,
```
Think this file can be entirely deleted

dibarbet commented on src/VisualStudio/LiveShare/Impl/RunCodeActionsHandler.cs:
```diff
@@ -7,127 +7,74 @@
     /// </summary>
-    internal abstract class RunCodeActionsHandler : ILspRequestHandler<LSP.ExecuteCommandParams, object, Solution>
+    [ExportExecuteWorkspaceCommand(CodeActionsHandler.RunCodeActionCommandName)]
+    internal class RunCodeActionsHandler : IExecuteWorkspaceCommandHandler
```
This should be moved from the liveshare layer down to where the rest of the handlers are

dibarbet commented on src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs:
```diff
@@ -62,15 +101,31 @@ private static LSP.CodeActionParams CreateCodeActionParams(LSP.Location caret)
+                Diagnostics = diagnostics,
             };
+
+            if (edit != null)
```
can't we just do action.Edit = edit?  if it's null it's no different from the default value

dibarbet commented on src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs:
```diff
@@ -62,15 +101,31 @@ private static LSP.CodeActionParams CreateCodeActionParams(LSP.Location caret)
+                action.Edit = edit;
+            }
+
+            if (command != null)
```
same as above

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -96,20 +135,60 @@ public CodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringServic
             return codeActions;
         }
+
+        internal static async Task<IEnumerable<KeyValuePair<CodeAction, CodeActionKind>>> GetCodeActionsAndKindAsync(
```
nit - I think this code would make more sense to me if we just iterate separately through the code fixes and refactorings, convert each list to LSP code action, then concat everything at the end rather than having to deal with the tuple stuff.

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,279 @@
+                    // is fixed.
+                    foreach (var docId in changedAdditionalDocuments)
+                    {
+                        var newDoc = applyChangesOperation.ChangedSolution.GetAdditionalDocument(docId);
```
these three loops are all doing basically the same thing, could we combine them?  Maybe a generic GetTextDocumentEdits(IEnumberable<T> documents, Func<DocumentId, T> GetChangedDocFunc)

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,279 @@
+                        pc => pc.GetAddedDocuments().Concat(pc.GetAddedAdditionalDocuments().Concat(pc.GetAddedAnalyzerConfigDocuments())));
+                    if (addedDocuments.Any())
+                    {
+                        return textDocumentEdits.ToArray();
```
It seems like if there are already textDocumentEdits in the array, then the check on line 105 will execute and will never hit the command.  Should this return an empty array?

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,279 @@
+            return codeAction;
+
+            // Local functions
+            static async Task<TextDocumentEdit[]> ComputeWorkspaceEdits(
```
I think it would be cleaner for this to be something like TryComputeWorkspaceEdits that returns false when unable to compute them and a command has to be used instead.

allisonchou replied:
Can that be done when the function is async? If we return a bool, then we would need to have an out parameter for the workspace edits, but I don't think we can use out parameters in async methods. I guess we could also return a tuple of (bool, WorkspaceEdits) but that seems a bit messy.

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,54 +38,90 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+                    Kind = codeActionKind,
+                    Diagnostics = request.Context.Diagnostics,
+                    Children = nestedActions.ToArray(),
+                    Data = new CodeActionResolveData { CodeActionParams = request, DistinctTitle = parentTitle + codeAction.Title }
```
are there cases where the nested code action titles conflict between two different nested groups?  If so, I wonder if we could use the codeactionprovider type name + title instead of having to specify the entire tree (unless the title can be the same within a nested group)

we should also doc on the CodeActionResolveData what goes inside the DistinctTitle

allisonchou replied:
Hmm, I think this would fail with 'Suppress or Configure issues' if we had two diagnostics that we could suppress/configure, e.g. we would have "ConfigureSeverityCodeFixProvider" + "None" twice in this case.

allisonchou commented:
@dibarbet I think I've addressed all your concerns via code changes or by responding to the initial feedback.
Just want to confirm with Manish that the resource string can be moved. Also, I'm not completely sure but I think a bug may need to be filed to address your feedback here: https://github.com/dotnet/roslyn/pull/45572#discussion_r448052958.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveData.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions
+{
+    /// <summary>
+    /// Provides data needed to resolve code actions.
```
not a helpful comment :)

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveData.cs:
```diff
@@ -0,0 +1,28 @@
+        /// have the same unique title.
+        /// </summary>
+        /// <remarks>
+        /// The distinct tiel is currently set as:
```
```suggestion
        /// The distinct title is currently set as:
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveData.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public LSP.Range Range { get; set; }
+
+        public LSP.TextDocumentIdentifier TextDocument { get; set; }
```
do these all need to be mutable props?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveData.cs:
```diff
@@ -0,0 +1,28 @@
+    internal class CodeActionResolveData
+    {
+        /// <summary>
+        /// The unique title of a code action. No two code actions should
```
i wouldn't call this a 'title' then.  'title' made me think this is just the "Generate Method" portion of the code action.  This really is a UniqueIdentifier sort of thing.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+        }
+
+        public override async Task<LSP.VSCodeAction> HandleRequestAsync(
+            LSP.VSCodeAction codeAction,
```
i'm a little confused as to the semantics here.  it takes in a code action... and returns a code action?  can you clarify the concept here?

allisonchou replied:
Per your suggestion I've added some details in the original post of this PR, as well as added more comments in the code.
The `CodeActionsResolveHandler` fills out the VSCodeAction's Edit and/or Command properties. The handler is only invoked when the user requests the code action (e.g. by hovering), so it saves unnecessary computation.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+
+            if (codeActions == null || !codeActions.Any())
+            {
+                return codeAction;
```
doc why this is appropriate.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                return codeAction;
+            }
+
+            var codeActionToResolve = GetCodeActionToResolve(data.DistinctTitle, codeActions);
```
so to resolve a code action, we have to compute all the code actions again?  seems costly.  can/should we store/save things locally across calls to speed this up?

dibarbet replied:
One suggestion I would have is to do the caching in a follow-up pr.  This matches the current behavior in nexus (always recompute) so we're not regressing anything.

This pr is already fairly big, and caching I think will be a little complex and would prefer to review separately.

CyrusNajmabadi replied:
i am ok with that decision for this PR. :)

allisonchou replied:
Opened #45727 to track this.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                codeAction.Command = SetCommand(codeAction, data);
+            }
+
+            return codeAction;
```
the data flow here is really weird to read through.  lots of little state variables affecting flow.  i would restructure to be more like:

```
if this must be a command
   do all the logic for treating this like a command here
else
  do all teh logic for treating as an edit here
done
```

trying to fork out and then rejoin your data flow is just making this confusing imo.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+            var runAsCommand = false;
+
+            var applyChangesOperations = operations.Where(operation => operation is ApplyChangesOperation);
```
```suggestion
            var applyChangesOperations = operations.OfType<ApplyChangesOperation>();
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                CancellationToken cancellationToken)
+            {
+                using var _ = ArrayBuilder<TextDocumentEdit>.GetInstance(out var textDocumentEdits);
+                foreach (ApplyChangesOperation applyChangesOperation in applyChangesOperations)
```
```suggestion
                foreach (var applyChangesOperation in applyChangesOperations)
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+
+            // Local functions
+            static async Task<TextDocumentEdit[]> ComputeWorkspaceEdits(
+                IEnumerable<CodeActionOperation> applyChangesOperations,
```
```suggestion
                IEnumerable<ApplyChangesOperation> applyChangesOperations,
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                        pc => pc.GetAddedDocuments().Concat(pc.GetAddedAdditionalDocuments().Concat(pc.GetAddedAnalyzerConfigDocuments())));
+                    if (addedDocuments.Any())
+                    {
+                        return Array.Empty<TextDocumentEdit>();
```
return null instead?  or a bool indicating this failed?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                    await GetTextDocumentEdits(
+                        textDocumentEdits, applyChangesOperation, solution, changedDocuments,
+                        applyChangesOperation.ChangedSolution.GetDocument, solution.GetDocument,
+                        cancellationToken).ConfigureAwait(false);
```
you are populating a parameter, not returning a value.  in cases like this, i prefer `AddTextDocumentEdits`

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                    // We won't get any results until https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+                    // is fixed.
+                    await GetTextDocumentEdits(
+                        textDocumentEdits, applyChangesOperation, solution, changedDocuments,
```
this seems wrong.  you're passing in changedDocuments instead of changedAnalyzerConfigDocuments (same issue below).  This feels untested as i would have thought this just wouldn't work.  and/or would crash.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                        var oldDoc = getOldDocumentFunc(docId);
+                        if (oldDoc == null || newDoc == null)
+                        {
+                            continue;
```
this shoudl never happen.  it indicates a bug in the caller since we shoudl only be passing in things that changed, so we must have a before/after.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                        var textChanges = newText.GetTextChanges(oldText).ToList();
+
+                        var edits = textChanges.Select(tc => ProtocolConversions.TextChangeToTextEdit(tc, oldText)).ToArray();
+                        var documentIdentifier = new VersionedTextDocumentIdentifier() { Uri = newDoc.GetURI() };
+                        textDocumentEdits.Add(new TextDocumentEdit() { TextDocument = documentIdentifier, Edits = edits.ToArray() });
```
```suggestion
                        var documentIdentifier = new VersionedTextDocumentIdentifier { Uri = newDoc.GetURI() };
                        textDocumentEdits.Add(new TextDocumentEdit { TextDocument = documentIdentifier, Edits = edits.ToArray() });
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                CommandIdentifier = CodeActionsHandler.RunCodeActionCommandName,
+                Title = codeAction.Title,
+                Arguments = new object[] { data }
+            };
```
just inline this entire local func.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+            };
+        }
+
+        internal static CodeAction? GetCodeActionToResolve(string distinctTitle, IEnumerable<CodeAction> codeActions)
```
internal?

CyrusNajmabadi replied:
please don't pass IEnumerables around.  It is impossible to tell if these ar elazy, and if iterating over them multiple times will hav a high perf cost.  use ImmutableArray.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+            // Searching for the matching code action. We compare against the distinct title (e.g. "Suppress IDExxxxNone")
+            // instead of the regular title (e.g. "None") since there's a chance that multiple code actions may have
+            // the same regular title.
+            CodeAction? codeActionToResolve = null;
```
```suggestion
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+        internal static CodeAction? GetCodeActionToResolve(string distinctTitle, IEnumerable<CodeAction> codeActions)
+        {
+            // Searching for the matching code action. We compare against the distinct title (e.g. "Suppress IDExxxxNone")
+            // instead of the regular title (e.g. "None") since there's a chance that multiple code actions may have
```
"None"?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+            foreach (var c in codeActions)
+            {
+                var action = CheckForMatchingAction(c, distinctTitle, currentTitle: "");
+                if (action != null)
+                {
+                    codeActionToResolve = action;
+                    break;
+                }
```
```suggestion
                if (action != null)
                    return action;
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+                }
+            }
+
+            return codeActionToResolve;
```
```suggestion
            return null;
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,257 @@
+
+        private static CodeAction? CheckForMatchingAction(CodeAction codeAction, string goalTitle, string currentTitle)
+        {
+            if (currentTitle + codeAction.Title == goalTitle)
```
introduce local for `currentTitle + codeAction.Title`.  you compute it several times (including in a loop below).

CyrusNajmabadi replied:
in general, when appending stuff like this it's nice to have some sort of delimeter between things so that you don't end up with words all jammed together.  Insetad, you get a nice broken up sequence of pieces that make the hierarchy clear.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,57 +38,113 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+                return Array.Empty<VSCodeAction>();
+            }
+
+            var (codeFixCollections, codeRefactorings) = await GetCodeFixesAndRefactoringsAsync(
+                                document, _codeFixService, _codeRefactoringService,
+                                request.Range, cancellationToken).ConfigureAwait(false);
```
```suggestion
            var (codeFixCollections, codeRefactorings) = await GetCodeFixesAndRefactoringsAsync(
                document, _codeFixService, _codeRefactoringService,
                request.Range, cancellationToken).ConfigureAwait(false);
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,57 +38,113 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+
+            var suppressionActions = codeFixes.Where(
+                a => a.Action is AbstractConfigurationActionWithNestedActions &&
+                (a.Action as AbstractConfigurationActionWithNestedActions)?.IsBulkConfigurationAction == false);
```
yikes.  needs comments.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,57 +38,113 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+                results.Add(GenerateVSCodeAction(request, new CodeActionWithNestedActions(
+                    CodeFixesResources.Suppress_or_Configure_issues,
+                    configureSuppressActions.Select(a => a.Action).ToImmutableArray(), true), CodeActionKind.QuickFix));
+            }
```
i need @mavasani for this sort of stuff.

CyrusNajmabadi replied:
still applies.

allisonchou replied:
Opened #45726 to track moving this logic to a uniform location (as per our earlier discussion).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -96,20 +155,29 @@ public CodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringServic
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
+            var codeFixCollections = await codeFixService.GetFixesAsync(document, textSpan, true, cancellationToken).ConfigureAwait(false);
+            var codeRefactorings = await codeRefactoringService.GetRefactoringsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
```
this is on a server right?  or is this inside VS?  if inside a server, i woudl do these in parallel.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -96,20 +155,29 @@ public CodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringServic
+        {
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
+            var codeFixCollections = await codeFixService.GetFixesAsync(document, textSpan, true, cancellationToken).ConfigureAwait(false);
```
named param for `true`

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/RunCodeActionsHandler.cs:
```diff
@@ -0,0 +1,81 @@
+namespace Microsoft.VisualStudio.LanguageServices.LiveShare
+{
+    /// <summary>
+    /// Runs code actions as a command on the server.
```
i'm not sure what this means.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/RunCodeActionsHandler.cs:
```diff
@@ -0,0 +1,81 @@
+            }
+
+            var actionToRun = CodeActionResolveHandler.GetCodeActionToResolve(runRequest.DistinctTitle, codeActions);
+            if (actionToRun != null)
```
```suggestion
            if (actionToRun == null)
                return false;
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/Initialize/InitializeHandler.cs:
```diff
@@ -54,6 +57,7 @@ public Task<LSP.InitializeResult> HandleRequestAsync(LSP.InitializeParams reques
                     DocumentHighlightProvider = true,
                     ReferencesProvider = true,
                     ProjectContextProvider = true,
+                    ExecuteCommandProvider = new ExecuteCommandOptions(),
```
this feels very vague.  should this be ExecuteCodeActionCommandOptions?

allisonchou replied:
`ExecuteCommandOptions` corresponds to the general LSP handler for commands. I added the LSP prefix to this line for clarity, and also added some further details about this in the main post of the PR.

CyrusNajmabadi commented:
Done with initial pass.  Things seem to be overall sensible.  I don't have a really good understanding of the expected flow here though, esp. with complicated cases.  Would def appreciate an overview in the main post of the PR.

allisonchou commented:
@CyrusNajmabadi Thank you for the feedback! I believe I've addressed all of your comments. Would it be possible to receive another review whenever you have time?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:
```diff
@@ -0,0 +1,118 @@
+            LSP.Range selection,
+            CancellationToken cancellationToken)
+        {
+            if (document == null)
```
can when doc when/why this happens?  i'm surprised anyone would call this with a null doc.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:
```diff
@@ -0,0 +1,118 @@
+                cancellationToken).ConfigureAwait(false);
+
+            var codeActions = codeFixCollections.SelectMany(c => c.Fixes.Select(f => f.Action)).Concat(
+                    codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
```
is htis the ordering the host will show the items in?  if so, this seems to not follow the ordering we have in VS (where we sort/order based on priority and other factors).  Is that intentional?  Is that problematic?  For example, rename is a refactoring, and we expect it to be high pri and at the top of the list.  this looks liek this wouldn't support that.

CyrusNajmabadi replied:
we may need to extract out hte logic in SuggestedActionSoruce to fetch and order the actions properly and then use that from that location as well as this one.

allisonchou replied:
Opened #45726 to track this.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:
```diff
@@ -0,0 +1,118 @@
+            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
+
+            var codeFixCollections = Array.Empty<CodeFixCollection>().ToImmutableArray();
+            var codeRefactorings = Array.Empty<CodeRefactoring>().ToImmutableArray();
```
```suggestion
            var codeRefactorings = ImmutableArray<CodeRefactoring>.Empty;
```

(applies to other locations as well).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:
```diff
@@ -0,0 +1,118 @@
+                    document, textSpan, cancellationToken).ConfigureAwait(false));
+
+            await Task.WhenAll(codeFixCollectionsTask, codeRefactoringsTask).ConfigureAwait(false);
+            return (codeFixCollections, codeRefactorings);
```
let's chat tomorrow on a cleaner way to do this.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:
```diff
@@ -0,0 +1,118 @@
+                }
+            }
+
+            return null;
```
is this possible?  doesit in indicate a bug somewhere?  should we be assert/logging instead?

CyrusNajmabadi replied:
in other words, code like this scares me.  it makes me feel lke we'll have a customer saying "my code actions aren't behaving properly" and we'll never be able to track down why.

allisonchou replied:
I think you're right. I pushed a commit that modifies all callers of the method to catch this case.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:
```diff
@@ -0,0 +1,118 @@
+                }
+            }
+
+            return null;
```
same concern.

allisonchou replied:
I think returning null is valid in this particular method since the caller loops through the top-level code actions to check for a match. So we should definitely find a match in some instance of this method, but not necessarily the current instance (if that makes sense).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveData.cs:
```diff
@@ -0,0 +1,37 @@
+        /// The unique identifier is currently set as:
+        /// name of top level code action + '.' + name of nested code action + '.' + name of nested nested code action + etc.
+        /// e.g. 'Suppress or Configure issues.Suppress IDEXXXX.in Source'
+        /// </remarks>
```
👍

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,195 @@
+    /// Resolves a code action by filling out its Edit and/or Command property.
+    /// The handler is triggered only when a user hovers over a code action, which
+    /// saves us from having to compute unnecessary edits.
+    /// </summary>
```
👍 

you mgiht want to consider fleshing this out just a teeny bit more.  i.e. this allows the basic actions to be computed quickly, and then more complex data, like the real edits or commands, can be computed on demand as necessary (for example, for hovering/preview of an action).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,195 @@
+                data.UniqueIdentifier, codeActions.ToImmutableArray());
+
+            // We didn't find a matching action, so just return the action without an edit or command.
+            if (codeActionToResolve == null)
```
doc if this is a bug or expected behavior.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,195 @@
+            var operations = await codeActionToResolve.GetOperationsAsync(cancellationToken).ConfigureAwait(false);
+            if (operations.IsEmpty)
+            {
+                return codeAction;
```
seems strange.  but i suppose nothing forces an action to have operations.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,195 @@
+                };
+            }
+
+            return codeAction;
```
the code flow here is extremely weird.  let's chat tomorrow, i think this can be definitely simplified.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,195 @@
+                    var newDoc = getNewDocumentFunc(docId);
+                    var oldDoc = getOldDocumentFunc(docId);
+
+                    var oldText = await oldDoc!.GetTextAsync(cancellationToken).ConfigureAwait(false);
```
instead of hte `!` which will likel have hard to understand dumps, can we just do some Contract.ThrowIfNulls above here?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,195 @@
+                    var oldText = await oldDoc!.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    var newText = await newDoc!.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                    var textChanges = newText.GetTextChanges(oldText).ToList();
```
```suggestion
                    var textChanges = newText.GetTextChanges(oldText);
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,81 +39,107 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
-                                codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
+            // Special case (also dealt with specially in local Roslyn): 
+            // If we have configure/suppress code actions, combine them under one top-level code action.
+            var configureSuppressActions = codeFixes.Where(a => a.Action is AbstractConfigurationActionWithNestedActions);
```
```suggestion
            var configureSuppressActions = codeFixes.OfType<AbstractConfigurationActionWithNestedActions>();
```

allisonchou replied:
Will this work since we wanted to check the type of `a.Action` and not just `a`?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,81 +39,107 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
-            // Flatten out the nested codeactions.
-            var nestedCodeActions = codeActions.Where(c => c is CodeAction.CodeActionWithNestedActions nc && nc.IsInlinable).SelectMany(nc => nc.NestedCodeActions);
-            codeActions = codeActions.Where(c => !(c is CodeAction.CodeActionWithNestedActions)).Concat(nestedCodeActions);
+            return results.ToArray();
```
this also feels like it' reimplementing logic that must exist elsewhere.  that doesn't seem desirable.  i would far prefer to have a single location where we sort/order these things, and then use it from VS and LSP.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/RunCodeActionsHandler.cs:
```diff
@@ -0,0 +1,85 @@
+
+            var actionToRun = CodeActionHelpers.GetCodeActionToResolve(runRequest.UniqueIdentifier, codeActions.ToImmutableArray());
+            if (actionToRun == null)
+            {
```
does this indicate a bug?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/RunCodeActionsHandler.cs:
```diff
@@ -0,0 +1,85 @@
+            var codeActions = await CodeActionHelpers.GetCodeActionsAsync(document, _codeFixService, _codeRefactoringService,
+                runRequest.Range, cancellationToken).ConfigureAwait(false);
+            if (codeActions == null)
+            {
```
does this indicate a bug?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/RunCodeActionsHandler.cs:
```diff
@@ -0,0 +1,85 @@
+        {
+            var runRequest = ((JToken)request.Arguments.Single()).ToObject<CodeActionResolveData>();
+            var document = _solutionProvider.GetDocument(runRequest.TextDocument);
+            var codeActions = await CodeActionHelpers.GetCodeActionsAsync(document, _codeFixService, _codeRefactoringService,
+                runRequest.Range, cancellationToken).ConfigureAwait(false);
```
```suggestion
            var codeActions = await CodeActionHelpers.GetCodeActionsAsync(
                document, _codeFixService, _codeRefactoringService, runRequest.Range, cancellationToken).ConfigureAwait(false);
```

stylistic nit for consideration.  both of these split the line to two lines.  however, this version keeps teh parameters together, instead of splitting at a random point.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/RunCodeActionsHandler.cs:
```diff
@@ -0,0 +1,85 @@
+            {
+                // TODO - This UI thread dependency should be removed.
+                // https://github.com/dotnet/roslyn/projects/45#card-20619668
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```
do this prior to the loop?

CyrusNajmabadi commented:
Done with pass.  I have some moderate concerns, but nothing critically blocking.  We could create work items to track some things if they are too large for this PR.  i have a general deep concern that we're being too loosy goosy with just papering over things that seem like they would be bugs, and that that could lead to a very frustrating user experience where things don't work and users don't know why.

allisonchou commented:
Opened #45726, #45727, and #45623 to track future work items for LSP code actions on the Roslyn side.

dibarbet commented on src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionResolveTests.cs:
```diff
@@ -0,0 +1,155 @@
+                diagnostics: null,
+                edit: new LSP.WorkspaceEdit
+                {
+                    DocumentChanges = new TextDocumentEdit[]
```
let's add a helper to create the workspace edit from markup, uri, and location since this is repeated

dibarbet commented on src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs:
```diff
@@ -27,50 +31,99 @@ void M()
+                title: CSharpAnalyzersResources.Use_implicit_type,
+                kind: CodeActionKind.Refactor,
+                children: Array.Empty<LSP.VSCodeAction>(),
+                data: new CodeActionResolveData
```
can this use CreateCodeActionResolveData?

dibarbet commented on src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs:
```diff
@@ -27,50 +31,99 @@ void M()
+                title: string.Format(FeaturesResources.Introduce_constant_for_0, "1"),
+                kind: CodeActionKind.Refactor,
+                children: Array.Empty<LSP.VSCodeAction>(),
+                data: new CodeActionResolveData
```
also here, can this use CreateCodeActionResolveData

dibarbet commented on src/VisualStudio/LiveShare/Impl/Microsoft.VisualStudio.LanguageServices.LiveShare.csproj:
```diff
@@ -14,10 +14,7 @@
-                               Name="CSharpLspPackage"
-                               Class="Microsoft.VisualStudio.LanguageServices.LiveShare.Client.Debugging.CSharpLspPackage"
-                               AllowsBackgroundLoad="true"/>
+    <PkgDefPackageRegistration Include="{bb7e83f4-eaf6-456c-b140-f8c027a7ed8a}" Name="CSharpLspPackage" Class="Microsoft.VisualStudio.LanguageServices.LiveShare.Client.Debugging.CSharpLspPackage" AllowsBackgroundLoad="true" />
```
probably not necessary to include this change

allisonchou commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,81 +39,107 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+                CodeActionParams request,
+                CodeAction codeAction,
+                CodeActionKind codeActionKind,
+                string parentTitle = "")
```
Hmm, tried this out per your suggestion - it seems to be a bit problematic since using a string builder makes handling nested code actions difficult due to a combination of recursion + using the same StringBuilder, i.e. all the nested options are concatenated with each other.

allisonchou commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:
```diff
@@ -35,81 +39,107 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+
+                // Temporarily filter out suppress and configure code actions, as we'll later combine them under a top-level
+                // code action.
+                if (codeFix.Action is AbstractConfigurationActionWithNestedActions)
```
Yep, I discussed this with Cyrus and think this would be a good high-pri follow-up PR item (#45726).

allisonchou commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs:
```diff
@@ -0,0 +1,200 @@
+                    // Changed analyzer config documents
+                    // We won't get any results until https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+                    // is fixed.
+                    await AddTextDocumentEdits(
```
The issue with changed documents should now be fixed due to a change on the Liveshare side (hooray).
There's a bug on the Roslyn side when modifying editorconfig documents (#44331 is the PR fix), but it should hopefully be merged in soon.
Still pending is the work to support adding documents via code actions in LSP.

allisonchou commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:
```diff
@@ -0,0 +1,108 @@
+                () => codeRefactoringService.GetRefactoringsAsync(document, textSpan, cancellationToken));
+
+            await Task.WhenAll(codeFixCollectionsTask, codeRefactoringsTask).ConfigureAwait(false);
+            return (await codeFixCollectionsTask.ConfigureAwait(false), await codeRefactoringsTask.ConfigureAwait(false));
```
I discussed this with Cyrus recently and was told that it would probably be good to keep the WhenAll in here. I think the logic was that if both codeFixCollectionsTask and codeRefactoringsTask error or cancel, then using WhenAll allows us to retrieve both messages, while if we awaited them separately then we might only get one.

CyrusNajmabadi replied:
Allison is correct.  It's an extremely subtle thing and easy to make a mistake around.  However, we can lose exceptions if we kick off tasks and then await each one serially.  then WhenAll ensures we properly see all of this, and doesn't hurt anything in terms of perf.  i.e. the later awaits will be no-ops because they will precheck if the tasks are complete and will grab the values if so.

dibarbet replied:
Ah interesting, that behavior makes sense now that I think about it.  Thanks for the explanation!

allisonchou commented on src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:
```diff
@@ -0,0 +1,108 @@
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
+
+            var codeFixCollectionsTask = Task.Run(
```
Hmm, so not sure if I'm fully understanding correctly, but I believe we still do need the Task.Run so that we can assign the tasks' results to the variables we return after the WhenAll call is complete.

dibarbet replied:
GetFixesAsync returns a Task which you should be able to assign to the codeFixCollectionsTask variable and pass that to Task.WhenAll

allisonchou replied:
Oh, gotcha. Thanks!

allisonchou commented:
It looks like modifying existing documents is now supported after a [bug fix](https://devdiv.visualstudio.com/DevDiv/_git/b50f6d39-f810-47e7-8b91-f4496f97d97a/commit/310ceac8b951803d028c8d0dc0228e4f6ae17630) on the Liveshare side. Adding new documents is still not supported.
There's currently a bug that affects modifying .editorconfig documents on the Roslyn side, but this should be fixed once #44331 is merged in.

allisonchou commented:
@CyrusNajmabadi @dibarbet Would it be possible to get another look at this PR? I believe all feedback has been addressed.

Outstanding bugs:
- Adding/removing documents is not supported: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
- Modifying an editorconfig document will auto-generate two unnecessary files on the client side. I talked to Manish, and this bug should be fixed once this PR is merged in: #44331
- Code actions with underscores in their title are not formatted correctly: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1150812/

Currently, the future work items on the Roslyn side are:
- #45726 (high-pri): Make the logic between LSP/local Roslyn uniform - specifically, extract out code in SuggestedActionSource to be used in both
- #45727: Enable caching in code action logic
